larpandoracontent/LArTwoDReco/LArClusterAssociation/MvaLowEClusterMergingAlgorithm.cc

```diff
@@ -0,0 +1,750 @@
+template <typename T>
+StatusCode MvaLowEClusterMergingAlgorithm<T>::Run()
+{
+    ++m_event;
+    std::map<const Cluster*, ClusterVector> clustersToMerge;
+
+    for (const std::string &clusterListName : m_inputClusterListNames)
+    {
+        try
+        {
+            const ClusterList *pClusterList = nullptr;
+            PANDORA_RETURN_RESULT_IF_AND_IF(STATUS_CODE_SUCCESS, STATUS_CODE_NOT_INITIALIZED, !=, PandoraContentApi::GetList(*this, clusterListName, pClusterList));
+
+            if (!pClusterList || pClusterList->empty())
+            {
+                if (PandoraContentApi::GetSettings(*this)->ShouldDisplayAlgorithmInfo()) 
+                    std::cout << "ClusterMergingAlgorithm: unable to find cluster list " << clusterListName << std::endl;
+
+                continue;
+            }
+
+            PandoraContentApi::ParticleFlowObject::Metadata lowEClustersMetadata;
+            this->EdgeHitComparer(pClusterList, clusterListName);
+        
+
+	}
+        
+        catch (StatusCodeException &statusCodeException)
+        {
+            throw statusCodeException;
+        }
+        
+	if (!clustersToMerge.empty())
+        {
+	       	//PANDORA_RETURN_RESULT_IF(STATUS_CODE_SUCCESS, !=, PandoraContentApi::SaveList(*this, m_clustersToMerge, clustersToMerge));
+		std::cout << std::endl;
+        }
+
+	if (m_writeTree)
+        {
+            PANDORA_MONITORING_API(SaveTree(this->GetPandora(), m_treeName.c_str(), m_fileName.c_str(), "RECREATE"))
+         }
+
+    }
+
+    return STATUS_CODE_SUCCESS;
+}
+
+//--------------------------------------------------------------------------------------------//
+
+template <typename T>
+const MCParticle* MvaLowEClusterMergingAlgorithm<T>::GetMCForCluster(const Cluster *const cluster, std::map<const Cluster*,
+    const MCParticle*> &clusterToMCMap) const
+{
+    const MCParticle* clusterMC = nullptr;
+
+    if (clusterToMCMap.count(cluster) > 0)
+    {
+        clusterMC = clusterToMCMap.at(cluster);
+    }
+    else
+    {
+        try
+        {
+            clusterMC = MCParticleHelper::GetMainMCParticle(cluster);
+            clusterToMCMap[cluster] = clusterMC;
+        }
+        catch (StatusCodeException e)
+        {
+            std::cout << "Failed to get MC particle for cluster of " << cluster->GetOrderedCaloHitList().size()
+                      << " : " << e.ToString() << std::endl;
+        }
+    }
+
+    return clusterMC;
+}
+
+//---------------------------------------------------------------------------------------------//
+
+template <typename T>
+bool MvaLowEClusterMergingAlgorithm<T>::IsValidToUse(const Cluster *const cluster, std::map<const Cluster*, bool> &clusterIsUsed) const
+{
+
+   // if (!cluster->IsAvailable())
+   //     return false;
+    
+    if (cluster->GetNCaloHits() < m_minNCaloHits)
+        return false;
+
+    if (cluster->GetNCaloHits() > m_upperHitThreshold)
+        return false;
+
+    if (clusterIsUsed.count(cluster) > 0)
```

Comment:
I'd put this first, no point in the other checks if the cluster is already in use